src/mono/mono/mini/method-to-ir.c

```diff
@@ -9546,6 +9558,70 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 			}
 
+			// ASSUME
+			//   interface ISomeIface { void Method(); }
+			//   struct SomeStruct : ISomeIface {...}
+			// OPTIMIZE
+			//   box SomeStruct
+			//   callvirt instance ISomeIface::Method()
+			// TO
+			//   call SomeStruct::Method()
+			guchar* callvirt_ip;
+			guint32 callvirt_proc_token;
+			if (!((cfg->compile_aot || cfg->compile_llvm) && !mono_class_is_def(klass)) && // we cannot devirtualize in AOT when using generics
```

Comment:
I don't understand this condition - together with the comment.    we don't often see `mono_class_is_def` used. `is_def` will be just the non-generic class/struct/interface types - it will exclude arrays, generic definitions and generic instances.  it's really a property of the metadata encoding, not usually something that mini cares about.

does the issue have to do with generic sharing?

Reply:
I was getting runtime errors under LLVM in the test `JIT/opt/Devirtualization/box1`. The test calls a method of a generic struct through an interface. The test passes with JIT. I noticed that calling methods of generic value types is more involved in AOT, so I disabled the optimization on LLVM if the class is generic. 

Maybe checking for gsharedvt instead would suffice?

Reply:
Yea - I would check for gsharedvt.  

We can make a follow-up work item to revisit the optimization in gsharedvt contexts.

Reply:
There are a lot of cases here so I would suggest restricting the set of allowed cases to a simple set right now so no generics, etc.